File: frontend/src/routes/companies/suppliers.js
/*************** Modules ***************/
import express from 'express';

import { create_params } from '../../utils/params.js';
import { get_data, send_data } from '../../utils/api.js';
import { require_auth, at_least_company } from '../../middlewares/auth.js';

/*************** Variables ***************/
const router = express.Router({ mergeParams: true });

/*************** Main route ***************/
router.get("/", require_auth, at_least_company, async (req, res) => {
    // Variables
    const permissions = req.permissions;
    
    const page = parseInt(req.query.page || "1", 10);
    const current_page = Math.max(page, 1);

    const type_of = req.query.type;
    const q = req.query.q;
    
    let suppliers = [];

    // Check permissions
    if (!permissions.includes("company.suppliers.read")) return res.redirect("/system-alert/403");

    // Request
    const params = create_params({ page: current_page, type_of: type_of, q: q })
    const response = await get_data(`/company/suppliers${params}`, {}, req);

    if (response.error) return res.redirect("/");

    const data = response.data;

    suppliers = data.suppliers

    // Render content
    return res.render("companies/suppliers/main", {
        suppliers,

        pagination: data.pagination,
        page: current_page,
        query: req.query
    });
});

/*************** Create Supplier - GET ***************/
router.get("/create", require_auth, at_least_company, async (req, res) => {
    // Variables
    const permissions = req.permissions;
    
    // Check permissions
    if (!permissions.includes("company.suppliers.create")) return res.redirect("/system-alert/403");

    // Render content
    return res.render("companies/suppliers/create");
});

/*************** Create Supplier - POST ***************/
router.post("/create", require_auth, at_least_company, async (req, res) => {
    // Variables
    const company_id = req.company_id;
    const permissions = req.permissions;

    let reason_name_v = "none";
    let document_v = "none";
    let email_v = "none";
    let phone_v = "none";
    let domain_v = "none";
    let address_v = "none";
    let supplier_type_v = "none";
    let is_active_v = "0";
    
    // Check permissions
    if (!permissions.includes("company.suppliers.create")) return res.redirect("/system-alert/403");

    // Body
    const { name, reason_name, document, email, phone, domain, address, supplier_type, is_active } = req.body;

    if (reason_name) reason_name_v = reason_name;
    if (document) document_v = document;
    if (email) email_v = email;
    if (phone) phone_v = phone;
    if (domain) domain_v = domain;
    if (address) address_v = address;
    if (supplier_type) supplier_type_v = supplier_type;
    if (is_active == "on") is_active_v = "1";

    const response = await send_data("/company/suppliers/create", {}, {
        name,

        reason_name: reason_name_v,
        document: document_v,
        email: email_v,
        phone: phone_v,
        domain: domain_v,
        address: address_v,
        supplier_type: supplier_type_v,
        is_active: is_active_v
    }, req);

    // Render content
    return res.redirect(`/companies/${company_id}/suppliers`)
});

/*************** Export ***************/
export default router

File: backend/routes/companies/suppliers.py
########## Modules ##########
from fastapi import APIRouter, Request, Depends

from sqlalchemy import or_, desc
from sqlalchemy.orm import Session

from db.database import get_db
from db.model import Supplier_Type, Supplier

from core.i18n import translate
from core.generator import get_uuid
from core.db_management import add_db
from core.responses import custom_response
from core.permissions import check_permissions
from core.validators import read_json_body, validate_required_fields
from core.utils import is_int, pagination, normalize_search

########## Variables ##########
router = APIRouter()

########## Create Product ##########
@router.get("/")
async def get_supplies(request: Request, page = 1, type_of = "all", q = None, db: Session = Depends(get_db)):
    ### Variables ###
    lang = request.state.lang
    user = request.state.user
    company_id = request.state.company_id

    suppliers = []

    ### Params ###
    page = is_int(page)

    if not page:
        page = 1

    limit = 12
    page = max(page, 1)
    offset = (page - 1) * limit

    if type_of != "all":
        if not type_of in Supplier_Type._value2member_map_:
            type_of = "all"

    if q:
        q = q.strip()
        
        if len(q) < 2:
            q = None

    search = normalize_search(q)

    ### Validation ###
    if user == None:
        return custom_response(status_code=400, message=translate(lang, "validation.require_auth"))

    ### Check permissions ###
    access, message = check_permissions(db, request, "company.suppliers.read", company_id)
    
    if not access:
        return custom_response(status_code=400, message=message)
    
    ### Filters ###
    filters = [
        Supplier.company_id == company_id
    ]

    if type_of != "all":
        filters.append(Supplier.type == Supplier_Type(type_of))

    if search:
        filters.append(
            or_(
                Supplier.name.ilike(f"%{search}%"),
                Supplier.reason_name.ilike(f"%{search}%"),
                Supplier.document.ilike(f"%{search}%")
            )
        )

    total_supliers = db.query(Supplier).filter(Supplier.company_id == company_id).count()

    suppliers_data = db.query(Supplier).filter(*filters).order_by(
        desc(Supplier.date)
    ).limit(limit).offset(offset).all()

    for supplier in suppliers_data:
        suppliers.append({
            "id": supplier.id,
            "name": supplier.name,
            "reason_name": supplier.reason_name,
            "document": supplier.document,
            "type": supplier.type,
            "is_active": supplier.is_active
        })

    return custom_response(status_code=200, message=translate(lang, "company.suppliers.create.success"), data={
        "suppliers": suppliers,
        "pagination": pagination(total_supliers, limit, offset)
    })


########## Create Supplier ##########
@router.post("/create")
async def create_supplier(request: Request, db: Session = Depends(get_db)):
    ### Variables ###
    lang = request.state.lang
    user = request.state.user
    company_id = request.state.company_id

    ### Validation ###
    if user == None:
        return custom_response(status_code=400, message=translate(lang, "validation.require_auth"))

    ### Check permissions ###
    access, message = check_permissions(db, request, "company.suppliers.create", company_id)
    
    if not access:
        return custom_response(status_code=400, message=message)
    
    ### Get Body ###
    supplier_check, error = await read_json_body(request)
    if error: 
        return custom_response(status_code=400, message=error)

    required_fields, error = validate_required_fields(supplier_check, [
        "name", "reason_name", "document", "email", "phone", "address", "supplier_type", "is_active"
    ])

    if error:
        return custom_response(status_code=400, message=translate(lang, "validation.required_f"), details=required_fields)
    
    if not supplier_check.supplier_type in Supplier_Type._value2member_map_:
        return custom_response(status_code=400, message=translate(lang, "company.suppliers.create.error.incorrect_type"))

    # Create Supplier
    new_supplier = Supplier(
        id = get_uuid(db, Supplier),
        name = supplier_check.name,
        type = Supplier_Type(supplier_check.supplier_type),

        company_id = company_id
    )

    reason_name = supplier_check.reason_name
    document = supplier_check.document
    email = supplier_check.email
    phone = supplier_check.phone
    address = supplier_check.address
    domain = supplier_check.domain
    is_active = supplier_check.is_active

    if reason_name != "none":
        new_supplier.reason_name = reason_name

    if document != "none":
        new_supplier.document = document

    if email != "none":
        new_supplier.email = email

    if phone != "none":
        new_supplier.phone = phone

    if address != "none":
        new_supplier.address = address
    
    if domain != "none":
        new_supplier.domain = domain

    if is_active == "0":
        new_supplier.is_active = False

    add_db(db, new_supplier)

    return custom_response(status_code=200, message=translate(lang, "company.suppliers.create.success"), data={
        "supplier_id": new_supplier.id
    })
